### PR TITLE
Enforce strict field initialization

### PR DESCRIPTION
**Pull request by dcodeIO**
As a follow-up to https://github.com/AssemblyScript/assemblyscript/pull/1155, this PR implements strict field initialization but makes it the default, so we always get the safety guarantees we need in AOT. I expect this to be somewhat controversial because it's stricter than TS in potentially unsafe scenarios.

To give a typical example:

```ts
class Foo {
  bar: Bar;
  constructor() {
    this.bar = new Bar(this);
    // illegal because ^ `Bar#constructor` can do anything with
    // the partially initialized `this` before `.bar` is assigned
  }
}
class Bar {
  constructor(public foo: Foo) {
  }
}
```

must instead be written like

```ts
class Foo {
  private _bar: Bar | null; // initialized with zero
  constructor() {
    this._bar = new Bar(this);
  }
  get bar() {
    return assert(this._bar);
  }
}
...
```

essentially enforcing an actual safety check for the case where `Foo#bar` is being accessed too early in `Bar#constructor`.

What this PR doesn't do compared to TS, however, is to enforce strict initialization of fields that can be trivially initialized with zero (which is our `undefined`), like of type `i32`, `f64` or a nullable type like `string | null`, so one can leave these uninitialized instead of adding boilerplate assignments, similar to locals or globals.

Still some work to be done, like object literals are currently failing, and it's likely that I missed some edge cases. Decided to put it here already so you can experience first hand how surprisingly breaking this change actually is.

For more examples of what's allowed and what's not, see:
* [Test cases](https://github.com/AssemblyScript/assemblyscript/blob/field-init/tests/compiler/field-initialization.ts)
* [Error test cases](https://github.com/AssemblyScript/assemblyscript/blob/field-init/tests/compiler/field-initialization-errors.ts)

- [x] I've read the contributing guidelines
- [x] Fix object literals

fixes https://github.com/AssemblyScript/assemblyscript/pull/1155

### CODE REVIEW

**Comment by dcodeIO** [comment]
With the last commit the manual assertion now becomes generated by the compiler when using definite assignment syntax on the field, so

```ts
class Foo {
  private bar!: Bar;
  constructor() {
    this.bar = new Bar(this); // valid because `.bar` is ignored
  }
}
...
```

effectively gets us what we want by ignoring `.bar` on field initialization checks. Whenever `.bar` is accessed, an assert is executed behind the scenes. Unlike TS, which doesn't specifically check, one gets the assertion at runtime ofc, so I also implemented a warning for the case where definite assignment syntax is unnecessary (which I guess can easily happen), so a user can remove it to avoid the runtime check.

File: src/compiler.ts [hunk 1]
```diff
@@ -9274,147 +9312,208 @@ export class Compiler extends DiagnosticEmitter {
...
-    flow.freeScopedLocals();
-    this.currentFlow = previousFlow;
+      if (baseClass) {
+        let parameterTypes = signature.parameterTypes;
+        let numParameters = parameterTypes.length;
+        let operands = new Array<ExpressionRef>(1 + numParameters);
+        operands[0] = module.local_get(0, nativeSizeType);
+        for (let i = 0; i < numParameters; ++i) {
+          operands[i + 1] = module.local_get(i + 1, parameterTypes[i].toNativeType());
+        }
```
Comment (MaxGraey) [inline src/compiler.ts:9425]:
```suggestion
        for (let i = 1; i < numParameters; ++i) {
          operands[i] = module.local_get(i, parameterTypes[i - 1].toNativeType());
        }
```

File: src/flow.ts [hunk 2]
```diff
@@ -537,6 +564,62 @@ export class Flow {
...
+        }
+      }
+    }
+  }
+
+  /** Tests if the specified `this` field has the specified flag or flags. */
+  isThisFieldFlag(field: Field, flag: FieldFlags): bool {
+    var fieldFlags = this.thisFieldFlags;
+    if (fieldFlags) {
+      return (changetype<FieldFlags>(fieldFlags.get(field)) & flag) == flag;
```
Comment (MaxGraey) [inline src/flow.ts:602]:
```suggestion
      return (changetype<FieldFlags>(fieldFlags).is(flag);
```

File: src/flow.ts [hunk 3]
```diff
@@ -829,6 +918,23 @@ export class Flow {
...
+    var leftFieldFlags = left.thisFieldFlags;
+    if (leftFieldFlags) {
+      let newFieldFlags = new Map<Field,FieldFlags>();
+      let rightFieldFlags = assert(right.thisFieldFlags);
+      for (let _keys = Map_keys(leftFieldFlags), i = 0, k = _keys.length; i < k; ++i) {
+        let key = _keys[i];
+        let leftFlags = changetype<FieldFlags>(leftFieldFlags.get(key));
+        if ((leftFlags & FieldFlags.INITIALIZED) != 0 && rightFieldFlags.has(key) && (changetype<FieldFlags>(rightFieldFlags.get(key)) & FieldFlags.INITIALIZED)) {
+          newFieldFlags.set(key, FieldFlags.INITIALIZED);
+        }
```
Comment (MaxGraey) [inline src/flow.ts:932]:
```suggestion
        if (
          (leftFlags & FieldFlags.INITIALIZED) != 0 && rightFieldFlags.has(key) && 
          (changetype<FieldFlags>(rightFieldFlags.get(key)) & FieldFlags.INITIALIZED)
        ) {
          newFieldFlags.set(key, FieldFlags.INITIALIZED);
        }
```

File: src/util/collections.ts [hunk 4]
```diff
@@ -3,30 +3,8 @@
...
-      cloned.add(v);
-    }
-    return cloned;
-  }
-  return new Set<V>();
-}
-
-export function makeMap<K,V>(original: Map<K,V> | null = null, overrides: Map<K,V> | null = null): Map<K,V> {
+/** Makes a unique map. Typically used to track contextual type arguemnts. */
+export function uniqueMap<K,V>(original: Map<K,V> | null = null, overrides: Map<K,V> | null = null): Map<K,V> {
```
Comment (MaxGraey) [inline src/util/collections.ts:7]: nice!

File: src/compiler.ts [hunk 5]
```diff
@@ -9420,8 +9420,8 @@ export class Compiler extends DiagnosticEmitter {
         let numParameters = parameterTypes.length;
         let operands = new Array<ExpressionRef>(1 + numParameters);
         operands[0] = module.local_get(0, nativeSizeType);
-        for (let i = 0; i < numParameters; ++i) {
```
Comment (MaxGraey) [inline src/compiler.ts]: Sorry forget mention that changes also required change exit condition from `i < numParameters` 
to `i <= numParameters`

**Comment by jedisct1** [comment]
Looks like if a constructor has the `@inline` attributes, fields it initializes are considered as not being initialized.

Is that intentional?

**Comment by dcodeIO** [comment]
Sounds like a bug to me. Need to check.
